src/shims/fs.rs

```diff
@@ -160,21 +172,33 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             return Ok(0);
         }
         let fd = this.read_scalar(fd_op)?.to_i32()?;
-        let buf_scalar = this.read_scalar(buf_op)?.not_undef()?;
-
-        // Remove the file handle to avoid borrowing issues.
-        this.remove_handle_and(fd, |mut handle, this| {
-            // Don't use `?` to avoid returning before reinserting the handle.
-            let bytes = this.force_ptr(buf_scalar).and_then(|buf| {
-                // FIXME: Don't use raw methods
-                this.memory
-                    .get_raw_mut(buf.alloc_id)?
-                    .get_bytes_mut(&*this.tcx, buf, Size::from_bytes(count))
-                    .map(|buffer| handle.file.read(buffer))
-            });
-            this.machine.file_handler.handles.insert(fd, handle).unwrap_none();
-            this.try_unwrap_io_result(bytes?.map(|bytes| bytes as i64))
-        })
+        let buf = this.read_scalar(buf_op)?.not_undef()?;
+
+        if let Some(handle) = this.machine.file_handler.handles.get_mut(&fd) {
+            let count = isize::try_from(count)
+                .map_err(|_| err_unsup_format!("Program tries to read into buffer too big for this host platform"))?;
+            // We want to read at most `count` bytes. We are sure that `count` is not negative
+            // because it was a target's `usize`. Also we are sure that its smaller than
+            // `usize::max_value()` because it is a host's `isize`.
+            let mut bytes = vec![0; count as usize];
+            let result = handle.file.read(&mut bytes);
+
+            match result {
+                Ok(c) => {
+                    let read_bytes = i64::try_from(c)
+                        .map_err(|_| err_unsup_format!("Number of read bytes {} cannot be transformed to i64", c))?;
```

Comment:
The result is no larger than `count`, and I think we can assume that `isize` and be cast to `i64` loslessly (we already assume that in many places).

So please `unwrap` or `expect` here, no need to forward an error to the user.